manifests/init.pp

```diff
@@ -7,3 +7,85 @@
   # include dns::config
   # include dns::service
 }
+
+# == Definition: dns_reverse_ptr_record
+#
+# This is a private type defined to allow the dns::record::a
+# defined resource type to create reverse ptr records for hosts
+# with multiple ip addresses.
+#
+# === Parameters
+#
+# * `$ip`
+# The ip address for the ptr record.  Defaults to the resource title.
+#
+# * `$host`
+# The (unqualified) host pointed to by this ptr record.
+#
+# * `$zone`
+# The domain of the host pointed to by this ptr record.
+#
+# === Actions
+#
+# Reformats the $ip, $host, $zone parameters to create a `dns::record::ptr` 
+# resource
+#
+# === Examples
+#
+# @example
+#     dns_reverse_ptr_record { '192.168.128.42':
+#       host => 'server' ,
+#       zone => 'example.com' ,
+#     }
+#
+# turns into:
+#
+# @example
+#     dns::record::ptr { '42.128.168.192.IN-ADDR.ARPA':
+#       host => '42' ,
+#       zone => '128.168.192.IN-ADDR.ARPA' ,
+#       data => 'server.example.com' ,
+#     }
+#
+# ---
+# @example
+#     dns_reverse_ptr_record { [ '192.168.128.68', '192.168.128.69', '192.168.128.70' ]:
+#       host => 'multihomed-server' ,
+#       zone => 'example.com' ,
+#     }
+#
+# turns into:
+#
+# @example
+#     dns::record::ptr { '68.128.168.192.IN-ADDR.ARPA':
+#       host => '68' ,
+#       zone => '128.168.192.IN-ADDR.ARPA' ,
+#       data => 'multihomed-server.example.com' ,
+#
+# @example
+#     dns::record::ptr { '69.128.168.192.IN-ADDR.ARPA':
+#       host => '69' ,
+#       zone => '128.168.192.IN-ADDR.ARPA' ,
+#       data => 'multihomed-server.example.com' ,
+#
+# @example
+#     dns::record::ptr { '70.128.168.192.IN-ADDR.ARPA':
+#       host => '70' ,
+#       zone => '128.168.192.IN-ADDR.ARPA' ,
+#       data => 'multihomed-server.example.com' ,
+#     }
+
+define dns_reverse_ptr_record (
```

Comment:
It works only because the init class is parsed first. Otherwise puppet would not be able to find this class because it is "hidden" in the init.pp. It violates the autoloading behavior:
https://docs.puppetlabs.com/puppet/latest/reference/lang_namespaces.html#autoloader-behavior

Can you link to the other pupetlabs modules that do this? I've never seen this before in a major module.

I don't think "private" is a much as a concern to me as being not-surprising. We could bump our version of stdlib and use the private keyword directly:
https://docs.puppetlabs.com/guides/style_guide.html#public-and-private

But yes please. Style-wise this is important to me. Both puppet and humans need to be able to find code in a non-surprising location. How about dns::record::reverse_ptr? (dns/manifests/record/reverse_ptr.pp)